syntax = "proto3";
option cc_generic_services = true;
import "common.proto";

package database; 

message DocumentHeader
{
   string key = 1;
   string category = 2;
}

message Document
{
   DocumentHeader header = 1;
   string body = 2;
}

enum QueryElementType
{
   Equals = 0;
   Less = 1;
   Greater = 2;
   Between = 3;
}

message QueryElement
{
   string key = 1;
   string value = 2;
   string value2 = 3;
   QueryElementType type = 4;
}

message DocumentQuery
{
   repeated QueryElement query = 1;
   string category = 2;
}

message Documents
{
   repeated Document result = 1;
}

message FullTextSearchResultEntry
{
   Document doc = 1;
   double match = 2;
}

message FullTextSearchResult
{
   repeated FullTextSearchResultEntry result = 1;
}

message FullTextSearchParameters
{
   string fulltext = 1;
}

service DatabaseService 
{
   rpc GetDocument (DocumentHeader) returns (Documents);
   rpc SearchDocuments (DocumentQuery) returns (Documents);
   rpc FullTextSearch (FullTextSearchParameters) returns (FullTextSearchResult);

   rpc AddDocument (Document) returns (common.UniqueId);
   rpc ModifyDocument (Document) returns (common.OperationResultMessage);
   rpc DeleteDocument (DocumentHeader) returns (common.OperationResultMessage);

   rpc Fetch(common.EmptyMessage) returns (Documents);
}